cmake/modules/hunter_download.cmake

```diff
@@ -142,9 +142,11 @@ function(hunter_download)
   set(HUNTER_GLOBAL_SCRIPT_DIR "${HUNTER_SELF}/scripts")
   set(HUNTER_PACKAGE_SCRIPT_DIR "${HUNTER_PACKAGE_SETUP_DIR}/scripts/")
   set(HUNTER_PACKAGE_HOME_DIR "${HUNTER_TOOLCHAIN_ID_PATH}/Build")
+  set(repository_hash "${HUNTER_${HUNTER_PACKAGE_NAME}_REPOSITORY_HASH}")
+  hunter_test_string_not_empty("${repository_hash}")
   set(
       HUNTER_PACKAGE_HOME_DIR
-      "${HUNTER_PACKAGE_HOME_DIR}/${HUNTER_PACKAGE_NAME}"
+      "${HUNTER_PACKAGE_HOME_DIR}/${HUNTER_PACKAGE_NAME}/${repository_hash}"
```

Comment:
> The other option it to read through the graph of dependencies. But this graph is not explicitly available, is it?

This graph is not static. You can't predict will your package (or it's next version) will depends on package Foo. First you have a code:

```cmake
if(MYPACKAGE_MORE_DEPS)
  hunter_add_package(Boo)
endif()
```

and next version is

```cmake
if(MYPACKAGE_MORE_DEPS)
  hunter_add_package(Boo)
  hunter_add_package(Foo) # new dependency!
endif()
```

There is a graph for cached packages, but not all packages are cacheable and graph can be used only after all packages requested by current project will be built.

As a general note I don't think the approach you're trying to implement is doable. You're trying to hack the really basic things Hunter relies on. I recommend you to think about creating Hunter archive on-the-fly either by [mixing several Hunter archives](https://github.com/ruslo/hunter/pull/384#issuecomment-213397810) or by [adding custom hunter.cmake recipies](https://github.com/ruslo/hunter/pull/295#issue-125450831).